src/Features/CSharp/Portable/UsePatternCombinators/CSharpUsePatternCombinatorsAnalyzer.cs

```diff
@@ -0,0 +1,224 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.UsePatternCombinators
+{
+    using static BinaryOperatorKind;
+    using static AnalyzedPattern;
+
+    internal static class CSharpUsePatternCombinatorsAnalyzer
+    {
+        public static bool Analyze(IOperation operation,
+            [NotNullWhen(true)] out AnalyzedPattern? pattern,
+            [NotNullWhen(true)] out ExpressionSyntax? target)
+        {
+            if ((pattern = ParsePattern(operation)) != null)
+                return TryGetTargetExpression(pattern, out target);
+            target = null;
+            return false;
+        }
+
+        private enum ConstantResult
+        {
+            /// <summary>
+            /// None of operands were constant.
+            /// </summary>
+            None,
+
+            /// <summary>
+            /// Signifies that the left operand is the constant.
+            /// </summary>
+            Left,
+
+            /// <summary>
+            /// Signifies that the right operand is the constant.
+            /// </summary>
+            Right,
+        }
+
+        private static AnalyzedPattern? ParsePattern(IOperation operation)
+        {
+            switch (operation)
+            {
+                case IBinaryOperation { OperatorKind: BinaryOperatorKind.Equals } op:
+                    return ParseConstantPattern(op);
+
+                case IBinaryOperation { OperatorKind: NotEquals } op:
+                    {
+                        var pattern = ParseConstantPattern(op);
+                        if (pattern == null)
+                            break;
+                        return Not.Create(pattern);
+                    }
+
+                case IBinaryOperation { OperatorKind: ConditionalOr, Syntax: BinaryExpressionSyntax syntax } op:
+                    return ParseBinaryPattern(op, isDisjunctive: true, syntax.OperatorToken);
+
+                case IBinaryOperation { OperatorKind: ConditionalAnd, Syntax: BinaryExpressionSyntax syntax } op:
+                    return ParseBinaryPattern(op, isDisjunctive: false, syntax.OperatorToken);
+
+                case IBinaryOperation op when IsRelationalOperator(op.OperatorKind):
+                    return ParseRelationalPattern(op);
+
+                case IUnaryOperation { OperatorKind: UnaryOperatorKind.Not } op:
+                    {
+                        var pattern = ParsePattern(op.Operand);
+                        if (pattern == null)
+                            break;
+                        return Not.Create(pattern);
+                    }
+
+                case IIsTypeOperation { Syntax: BinaryExpressionSyntax { Right: TypeSyntax type } } op:
+                    return new Type(type, op.ValueOperand);
+
+                case IIsPatternOperation { Pattern: { Syntax: PatternSyntax pattern } } op:
+                    return new Source(pattern, op.Value);
+
+                case IParenthesizedOperation op:
+                    return ParsePattern(op.Operand);
```

Comment:
when emitting, what ensures proper parenthesization?  does it fall out of the work i did previously for SyntaxGenerator and simplification?

Reply:
Yes, we call Parenthesize for nested nodes and the rest is magic.